README
NAME
    MooseX::FunkyAttributes - add code smell to your Moose attributes

SYNOPSIS
       package Circle;
   
       use Moose;
       use MooseX::FunkyAttributes;
   
       has radius => (
          is          => 'rw',
          isa         => 'Num',
          predicate   => 'has_radius',
       );
   
       has diameter => (
          traits      => [ FunkyAttribute ],
          is          => 'rw',
          isa         => 'Num',
          custom_get  => sub { 2 * $_->radius },
          custom_set  => sub { $_->radius( $_[-1] / 2 ) },
          custom_has  => sub { $_->has_radius },
       );

DESCRIPTION
    The MooseX::FunkyAttributes module itself just provides some convenience
    functions for the attribute traits that are distributed with it.

    The grand uniting idea behind the traits is that although Moose
    generally uses blessed hashrefs for object internals, storing each
    attribute as an entry in the hash, sometimes you may want to change that
    behaviour for individual attributes. For example, you could use the
    inside-out technique for one attribute to suppress it from appearing in
    Data::Dumper dumps.

    The traits bundled with MooseX::FunkyAttributes are as follows. Please
    see the documentation for each individual trait for further details.

    *   MooseX::FunkyAttributes::Role::Attribute

        Allows you to override the get, set, clear and predicate operations
        for an attribute. These are just the raw operations you're
        overriding; you do not need to implement type constraint checking,
        coercians, checking required attributes, defaults, builders, etc.

        With this trait you can create attributes which are calculated on
        the fly (as in the SYNOPSIS) or stored somewhere outside the
        object's blessed hashref.

    *   MooseX::FunkyAttributes::Role::Attribute::InsideOut

        This trait stores the attribute using the inside-out technique. If
        you want your whole object to be inside-out, then use
        MooseX::InsideOut.

    *   MooseX::FunkyAttributes::Role::Attribute::Delegated

        This trait delegates the storage of one attribute to another
        attribute. For example:

           package Head;
           use Moose;
           has mouth => (
              is           => 'ro',
              isa          => 'Mouth',
           );
   
           package Person;
           use Moose;
           use MooseX::FunkyAttributes;
           has head => (
              is           => 'ro',
              isa          => 'Head',
           );
           has mouth => (
              is           => 'ro',
              isa          => 'Mouth::Human',
              traits       => [ DelegatedAttribute ],
              delegated_to => 'head',
           );

        It is not dissimilar to the Moose's idea of "handles" (in the
        example above, we could have not defined a "mouth" attribute as part
        of the "Person" class, and just specified that "head" handles
        "mouth") however because "mouth" is now a proper attribute in the
        Person class (and will show up via Moose's introspection API), it
        can have its own type constraints, coercion, trigger, predicate,
        etc.

    The MooseX::FunkyAttributes module itself exports some constants (e.g.
    "FunkyAttribute" in the SYNOPSIS) which can be used as abbreviations for
    the full trait package names. These constants are:

    "FunkyAttribute"
    "InsideOutAttribute"
    "DelegatedAttribute"

BUGS
    Please report any bugs to
    <http://rt.cpan.org/Dist/Display.html?Queue=MooseX-FunkyAttributes>.

SEE ALSO
    MooseX::FunkyAttributes::Role::Attribute,
    MooseX::FunkyAttributes::Role::Attribute::InsideOut,
    MooseX::FunkyAttributes::Role::Attribute::Delegated.

    These effect storage for whole object instances; not just one attribute:
    MooseX::GlobRef, MooseX::InsideOut, MooseX::ArrayRef.

    MooseX::CustomInitArgs - if you have (as in the SYNOPSIS) one attribute
    which is calculated from another (diameter from radius),
    MooseX::CustomInitArgs will allow you to accept both attributes in the
    constructor (i.e. accept a diameter in the constructor, halve it, and
    set the radius attribute).

AUTHOR
    Toby Inkster <tobyink@cpan.org>.

COPYRIGHT AND LICENCE
    This software is copyright (c) 2012-2013 by Toby Inkster.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.

DISCLAIMER OF WARRANTIES
    THIS PACKAGE IS PROVIDED "AS IS" AND WITHOUT ANY EXPRESS OR IMPLIED
    WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED WARRANTIES OF
    MERCHANTIBILITY AND FITNESS FOR A PARTICULAR PURPOSE.



Changes
MooseX-FunkyAttributes
======================

Created:      2012-09-14
Home page:    <https://metacpan.org/release/MooseX-FunkyAttributes>
Bug tracker:  <http://rt.cpan.org/Dist/Display.html?Queue=MooseX-FunkyAttributes>
Maintainer:   Toby Inkster <mailto:tobyink@cpan.org>

0.002  2013-02-08

 - (Documentation) Document the metaobject interface (albeit briefly).
 - (Documentation) Document the test suite.
 - (Documentation) Include example scripts.
 - (Documentation) Show MooseX::CustomInitArgs in example, and include in
   SEE ALSO.
 - General tiding up.

0.001  2012-09-17  # Initial release




lib/MooseX/FunkyAttributes/Meta/Accessor.pm
package MooseX::FunkyAttributes::Meta::Accessor;

use 5.008;
use strict;
use warnings;

BEGIN {
	$MooseX::FunkyAttributes::Meta::Accessor::AUTHORITY = 'cpan:TOBYINK';
	$MooseX::FunkyAttributes::Meta::Accessor::VERSION   = '0.002';
}

use Moose;
use namespace::autoclean;

extends qw(Moose::Meta::Method::Accessor);

around _instance_is_inlinable => sub
{
	my $orig = shift;
	my $self = shift;
	my $attr = $self->associated_attribute;
	return $attr->accessor_should_be_inlined
		if $attr->does('MooseX::FunkyAttributes::Role::Attribute');
	$self->$orig(@_);
};

1;

__END__

=head1 NAME

MooseX::FunkyAttributes::Meta::Accessor - shim for inlining

=head1 DESCRIPTION

This is a small subclass of L<Moose::Meta::Method::Accessor> which defers
to the attribute the decision on whether the accessor should be inlined.

This is quite uninteresting.

=head1 BUGS

Please report any bugs to
L<http://rt.cpan.org/Dist/Display.html?Queue=MooseX-FunkyAttributes>.

=head1 SEE ALSO

L<MooseX::FunkyAttributes>, L<Moose::Meta::Method::Accessor>.

=head1 AUTHOR

Toby Inkster E<lt>tobyink@cpan.orgE<gt>.

=head1 COPYRIGHT AND LICENCE

This software is copyright (c) 2012-2013 by Toby Inkster.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=head1 DISCLAIMER OF WARRANTIES

THIS PACKAGE IS PROVIDED "AS IS" AND WITHOUT ANY EXPRESS OR IMPLIED
WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED WARRANTIES OF
MERCHANTIBILITY AND FITNESS FOR A PARTICULAR PURPOSE.



lib/MooseX/FunkyAttributes/Role/Attribute.pm
package MooseX::FunkyAttributes::Role::Attribute;

use 5.008;
use strict;
use warnings;

BEGIN {
	$MooseX::FunkyAttributes::Role::Attribute::AUTHORITY = 'cpan:TOBYINK';
	$MooseX::FunkyAttributes::Role::Attribute::VERSION   = '0.002';
}

use Moose::Role;

use aliased 'MooseX::FunkyAttributes::Meta::Accessor';
use namespace::autoclean;

has custom_get => (
	is         => 'ro',
	isa        => 'CodeRef',
	required   => 1,
);

has custom_set => (
	is         => 'ro',
	isa        => 'CodeRef',
	required   => 1,
);

has custom_has => (
	is         => 'ro',
	isa        => 'CodeRef',
	required   => 1,
);

has custom_clear => (
	is         => 'ro',
	isa        => 'CodeRef',
	predicate  => 'has_custom_clear',
);

has custom_weaken => (
	is         => 'ro',
	isa        => 'CodeRef',
	predicate  => 'has_custom_weaken',
);

has custom_init => (
	is         => 'ro',
	isa        => 'CodeRef',
	predicate  => 'has_custom_init',
);

my @i = qw( set get weaken has clear );
for my $i (@i)
{
	my $custom = "custom_inline_$i";
	has $custom => (
		is        => 'ro',
		isa       => 'CodeRef',
		predicate => "has_$custom",
	);
	
	my $guts_method =
		( $i =~ /^(weaken|clear)$/ )
			? "_inline_${i}_value"
			: "_inline_instance_${i}";
	
	around $guts_method => sub
	{
		my ($orig, $self, @args) = @_;
		if ($self->has_all_inliners) {
			return $self->$custom->($self, @args);
		}
		return $self->$orig(@args);
	};
}

around _inline_weaken_value => sub
{
	my ($orig, $self, @args) = @_;
	return unless $self->is_weak_ref;
	$self->$orig(@args);
};

has has_all_inliners => (
	is         => 'ro',
	isa        => 'Bool',
	lazy_build => 1,
);

sub _build_has_all_inliners
{
	my $self = shift;
	for (@i) {
		my $predicate = "has_custom_inline_$_";
		return unless $self->$predicate;
	}
	return 1;
}

sub accessor_should_be_inlined
{
	shift->has_all_inliners;
}

after _process_options => sub
{
	my ($class, $name, $options) = @_;
	
	if (defined $options->{clearer}
	and not defined $options->{custom_clear})
	{
		confess "can't set clearer without custom_clear";
	}

	if ($options->{weak_ref}
	and not defined $options->{custom_weaken})
	{
		confess "can't set weak_ref without custom_weaken";
	}
};

override accessor_metaclass => sub { Accessor };

override get_raw_value => sub
{
	my ($attr) = @_;
	local $_ = $_[1];
	return $attr->custom_get->(@_);
};

override set_raw_value => sub
{
	my ($attr) = @_;
	local $_ = $_[1];
	return $attr->custom_set->(@_);
};

override has_value => sub
{
	my ($attr) = @_;
	local $_ = $_[1];
	return $attr->custom_has->(@_);
};

override clear_value => sub
{
	my ($attr) = @_;
	local $_ = $_[1];
	return $attr->custom_clear->(@_);
};

override _weaken_value => sub
{
	my ($attr) = @_;
	local $_ = $_[1];
	return $attr->custom_weaken->(@_);
};

override set_initial_value => sub
{
	my ($attr) = @_;
	local $_ = $_[1];
	if ($attr->has_custom_init) {
		return $attr->custom_init->(@_);
	}
	return $attr->custom_set->(@_);
};

1;

__END__

=head1 NAME

MooseX::FunkyAttributes::Role::Attribute - custom get/set/clear/has coderefs

=head1 SYNOPSIS

   package Circle;
   
   use Moose;
   use MooseX::FunkyAttributes;
   
   has radius => (
      is          => 'rw',
      isa         => 'Num',
      predicate   => 'has_radius',
   );
   
   has diameter => (
      traits      => [ FunkyAttribute ],
      is          => 'rw',
      isa         => 'Num',
      custom_get  => sub { 2 * $_->radius },
      custom_set  => sub { $_->radius( $_[-1] / 2 ) },
      custom_has  => sub { $_->has_radius },
   );

=head1 DESCRIPTION

This is the base trait which the other MooseX::FunkyAttribute traits inherit
from. It allows you to provide coderefs to handle the business of storing and
retrieving attribute values.

So instead of storing your attribute values in the object's blessed hashref,
you could calculate them on the fly, or store them in a file or database, or
an external hashref, or whatever.

=head2 Options

If your attribute uses this trait, then you I<must> provide at least the
following three coderefs:

=over

=item C<< custom_set => CODE ($meta, $instance, $value) >>

The code which implements setting an attribute value. Note that this code
does I<not> need to implement type constraint checks, etc. C<< $meta >> is a
L<Moose::Meta::Attribute> object describing the attribute; C<< $instance >>
is the object itself.

C<< $_ >> is available as an alias for the instance.

=item C<< custom_get => CODE ($meta, $instance) >>

The code which implements getting an attribute value. 

It should return the value.

=item C<< custom_has => CODE ($meta, $instance) >>

The code which implements the predicate functionality for an attribute. That
is, it should return true if the attribute has been set, and false if the
attribute is unset. (Note that Moose does allow attribute values to be set to
undefined, so settedness is not the same as definedness.)

=back

The following three additional coderefs are optional:

=over

=item C<< custom_clear => CODE ($meta, $instance) >>

The code which clears an attribute value, making it unset.

If you do not provide this, then your attribute cannot be cleared once set.

=item C<< custom_init => CODE ($meta, $instance, $value) >>

Like C<custom_set> but used during object construction.

If you do not provide this, then the C<custom_set> coderef will be used in its
place.

=item C<< custom_weaken => CODE ($meta, $instance) >>

The code which weakens an attribute value that is a reference.

If you do not provide this, then your attribute cannot be a weak ref.

=back

Moose attempts to create inlined attribute accessors whenever possible. The
following coderefs can be defined which must return strings of Perl code
suitable for inlining the accessors. They are each optional, but unless all
of them are defined, your attribute will not be inlined.

=over

=item C<< custom_inline_set => CODE ($meta, $instance_string, $value_string) >>

C<< $instance_string >> is a string representing the name of the instance
variable, such as C<< "\$self" >>. C<< $value_string >> is a string which
evaluates to the value.

An example for the C<diameter> example in the SYNOPSIS

   custom_inline_set => sub {
      my ($meta, $i, $v) = @_;
      return sprintf('%s->{radius} = (%s)/2', $i, $v);
   },

=item C<< custom_inline_get => CODE ($meta, $instance_string) >>

An example for the C<diameter> example in the SYNOPSIS

   custom_inline_get => sub {
      my ($meta, $i) = @_;
      return sprintf('%s->{radius} * 2', $i);
   },

=item C<< custom_inline_has => CODE ($meta, $instance_string) >>

An example for the C<diameter> example in the SYNOPSIS

   custom_inline_has => sub {
      my ($meta, $i) = @_;
      return sprintf('exists %s->{radius}', $i);
   },

=item C<< custom_inline_clear => CODE ($meta, $instance_string) >>

An example for the C<diameter> example in the SYNOPSIS

   custom_inline_has => sub {
      my ($meta, $i) = @_;
      return sprintf('delete %s->{radius}', $i);
   },

=item C<< custom_inline_weaken => CODE ($meta, $instance_string) >>

An example for the C<diameter> example in the SYNOPSIS

   custom_inline_has => sub {
      my ($meta, $i) = @_;
      return sprintf('Scalar::Util::weaken(%s->{radius})', $i);
   },

(Not that weakening a Num makes any sense...)

=back

Your attribute metaobject has the following methods (in addition to the
standard L<Moose::Meta::Attribute> stuff):

=over

=item C<custom_get>

=item C<custom_set>

=item C<custom_has>

=item C<custom_clear>, C<has_custom_clear>

=item C<custom_weaken>, C<has_custom_weaken>

=item C<custom_init>, C<has_custom_init>

=item C<custom_inline_get>, C<has_custom_inline_get>

=item C<custom_inline_set>, C<has_custom_inline_set>

=item C<custom_inline_has>, C<has_custom_inline_has>

=item C<custom_inline_clear>, C<has_custom_inline_clear>

=item C<custom_inline_weaken>, C<has_custom_inline_weaken>

=item C<accessor_should_be_inlined>

=item C<has_all_inliners>

=back

=head1 BUGS

Please report any bugs to
L<http://rt.cpan.org/Dist/Display.html?Queue=MooseX-FunkyAttributes>.

=head1 SEE ALSO

L<MooseX::FunkyAttributes>.

=head1 AUTHOR

Toby Inkster E<lt>tobyink@cpan.orgE<gt>.

=head1 COPYRIGHT AND LICENCE

This software is copyright (c) 2012-2013 by Toby Inkster.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=head1 DISCLAIMER OF WARRANTIES

THIS PACKAGE IS PROVIDED "AS IS" AND WITHOUT ANY EXPRESS OR IMPLIED
WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED WARRANTIES OF
MERCHANTIBILITY AND FITNESS FOR A PARTICULAR PURPOSE.



lib/MooseX/FunkyAttributes/Role/Attribute/Delegated.pm
package MooseX::FunkyAttributes::Role::Attribute::Delegated;

use 5.008;
use strict;
use warnings;

BEGIN {
	$MooseX::FunkyAttributes::Role::Attribute::Delegated::AUTHORITY = 'cpan:TOBYINK';
	$MooseX::FunkyAttributes::Role::Attribute::Delegated::VERSION   = '0.002';
}

use Moose::Role;
use namespace::autoclean;

with qw(MooseX::FunkyAttributes::Role::Attribute);

has delegated_to        => (is => 'ro', isa => 'Str', required => 1);
has delegated_accessor  => (is => 'ro', isa => 'Str');
has delegated_predicate => (is => 'ro', isa => 'Str');
has delegated_clearer   => (is => 'ro', isa => 'Str');

before _process_options => sub
{
	my ($class, $name, $options) = @_;
	
	my $to = $options->{delegated_to}
		or confess "Required option 'delegated_to' missing";
	
	# Meh... we should use Moose's introspection to get the name of accessors, clearers, etc.
	# ... actually we can't do that. We don't know at attribute creation time, what sort of
	# object $self->$to will be!!
	#
	
	$options->{custom_weaken}        ||= sub { 0   };  # :-(
	$options->{custom_inline_weaken} ||= sub { q() };  # :-(
	
	$options->{delegated_accessor} = (
		my $accessor  = exists $options->{delegated_accessor} ? $options->{delegated_accessor} : $name
	);
	my $private   = !!($accessor =~ /^_/);
		
	if ($accessor and not exists $options->{custom_get})
	{
		$options->{custom_get} = sub { $_[1]->$to->$accessor };
		$options->{custom_inline_get} ||= sub {
			my ($self, $inst, $val) = @_;
			qq( $inst->$to->$accessor() )
		};
	}
	
	if ($accessor and not exists $options->{custom_set})
	{
		$options->{custom_set} = sub { $_[1]->$to->$accessor($_[2]) };
		$options->{custom_inline_set} ||= sub {
			my ($self, $inst, $val) = @_;
			qq( $inst->$to->$accessor($val) )
		};
	}
	
	$options->{delegated_predicate} = (
		my $predicate = exists $options->{delegated_predicate} ? $options->{delegated_predicate} : ($private ? "_has$accessor"   : "has_$accessor")
	);
	
	if ($predicate and not exists $options->{custom_has})
	{
		$options->{custom_has} = sub { $_[1]->$to->$predicate };
		$options->{custom_inline_has} ||= sub {
			my ($self, $inst) = @_;
			qq( $inst->$to->$predicate() )
		};
	}
	
	$options->{delegated_clearer} = (
		my $clearer   = exists $options->{delegated_clearer}   ? $options->{delegated_clearer}   : ($private ? "_clear$accessor" : "clear_$accessor")
	);
	
	if ($clearer and not exists $options->{custom_clear})
	{
		$options->{custom_clear} = sub { $_[1]->$to->$clearer };
		$options->{custom_inline_clear} ||= sub {
			my ($self, $inst) = @_;
			qq( $inst->$to->$clearer() )
		};
	}
	
	delete $options->{$_} for
		grep { not defined $options->{$_} }
		grep { /^delegated_/ }
		keys %$options;
};

1;


__END__

=head1 NAME

MooseX::FunkyAttributes::Role::Attribute::Delegated - delegate an attribute to another object

=head1 SYNOPSIS

   package Head;
   
   use Moose;
   
   has mouth => (
      is           => 'ro',
      isa          => 'Mouth',
   );
   
   package Person;
   
   use Moose;
   use MooseX::FunkyAttributes;
   
   has head => (
      is           => 'ro',
      isa          => 'Head',
   );
   
   has mouth => (
      is           => 'ro',
      isa          => 'Mouth::Human',
      traits       => [ DelegatedAttribute ],
      delegated_to => 'head',
   );

=head1 DESCRIPTION

This trait delegates the storage of one attribute's value to the object stored
in another attribute. The example in the SYNOPSIS might have been written using
Moose's native delegation as:

   package Head;
   
   use Moose;
   
   has mouth => (
      is           => 'ro',
      isa          => 'Mouth',
   );
   
   package Person;
   
   use Moose;
   
   has head => (
      is           => 'ro',
      isa          => 'Head',
      handles      => [qw( mouth )],
   );

However, there are some differences. Using native delegation, C<mouth>
will be treated as a method using Moose's introspection API
(C<< Person->meta->get_all_methods >>) and not as an attribute
(C<< Person->meta->get_all_attributes >>). Using this API, C<mouth> is
a proper attribute of C<Person>; it just relies on the C<Head> object for
storage.

Because C<mouth> is a proper attribute of C<Person>, it can perform
delegations of its own; can have its own type constraints, etc.

   has mouth => (
      is           => 'ro',
      isa          => 'Mouth::Human',
      traits       => [ DelegatedAttribute ],
      delegated_to => 'head',
      handles      => [qw/ speak kiss vomit eat /], # but not necessarily
   );                                               # in that order

=head2 Options

=over

=item C<< delegated_to => STR >>

The name of the other attribute to delegate this attribute to. This is the
only required option.

=item C<< delegated_accessor => STR >>

This option may be used if you wish to rename the delegated attribute. For
example:

   package Person;
   
   has pie_hole => (
      is           => 'ro',
      isa          => 'Mouth::Human',
      traits       => [ DelegatedAttribute ],
      delegated_to => 'head',
      delegated_accessor => 'mouth',
   );

Now C<< $person->pie_hole >> is equivalent to C<< $person->head->mouth >>.

If omitted, then it is assumed that the attribute has the same name in both
classes. If explicitly set to C<undef>, then this assumption is not made, and
the accessor is unknown. If the accessor is unknown, then this trait gets
somewhat stuck, so you'd need to provide C<custom_get> and C<custom_set>
options (see L<MooseX::FunkyAttributes::Role::Attribute>).

=item C<< delegated_predicate => STR >>

Like C<delegated_accessor>, but for the attribute's predicate. If omitted,
assumes the convention of C<< has_$accessor >>. An explicit undef can
avoid this assumption, but then you'll need to provide C<custom_has>.

=item C<< delegated_clearer => STR >>

Like C<delegated_accessor>, but for the attribute's clearer. If omitted,
assumes the convention of C<< clear_$accessor >>. An explicit undef can
avoid this assumption, but then you'll need to provide C<custom_has> if
you want to define a clearer.

=back

All the C<custom_blah> and C<custom_inline_blah> options from
L<MooseX::FunkyAttributes::Role::Attribute> are also available. The
C<delegated_blah> options above are essentially just shortcuts
for defining them.

Your attribute metaobject has the following methods (in addition to the
standard L<MooseX::FunkyAttributes::Role::Attribute> and
L<Moose::Meta::Attribute> stuff):

=over

=item C<delegated_to>

=item C<delegated_accessor>

=item C<delegated_clearer>

=item C<delegated_predicate>

=back

=head1 BUGS

Please report any bugs to
L<http://rt.cpan.org/Dist/Display.html?Queue=MooseX-FunkyAttributes>.

=head1 SEE ALSO

L<MooseX::FunkyAttributes>.

=head1 AUTHOR

Toby Inkster E<lt>tobyink@cpan.orgE<gt>.

=head1 COPYRIGHT AND LICENCE

This software is copyright (c) 2012-2013 by Toby Inkster.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=head1 DISCLAIMER OF WARRANTIES

THIS PACKAGE IS PROVIDED "AS IS" AND WITHOUT ANY EXPRESS OR IMPLIED
WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED WARRANTIES OF
MERCHANTIBILITY AND FITNESS FOR A PARTICULAR PURPOSE.



lib/MooseX/FunkyAttributes/Role/Attribute/InsideOut.pm
package MooseX::FunkyAttributes::Role::Attribute::InsideOut;

use 5.008;
use strict;
use warnings;

BEGIN {
	$MooseX::FunkyAttributes::Role::Attribute::InsideOut::AUTHORITY = 'cpan:TOBYINK';
	$MooseX::FunkyAttributes::Role::Attribute::InsideOut::VERSION   = '0.002';
}

use Hash::FieldHash ();
use Scalar::Util ();
use Moose::Role;
use namespace::autoclean;

with qw(MooseX::FunkyAttributes::Role::Attribute);

our @_HASHES;

my $i = 0;
before _process_options => sub
{
	my ($class, $name, $options) = @_;
	
	my $hashcount = $i++;
	Hash::FieldHash::fieldhash my %h;
	$_HASHES[$hashcount] = \%h;
	
	$options->{custom_get}           = sub { $h{ $_[1] } };
	$options->{custom_set}           = sub { $h{ $_[1] } = $_[2] };
	$options->{custom_has}           = sub { exists $h{ $_[1] } };
	$options->{custom_clear}         = sub { delete $h{ $_[1] } };
	$options->{custom_weaken}        = sub { Scalar::Util::weaken( $h{ $_[1] } ) };
	$options->{custom_init}          = sub { $h{ $_[1] } = $_[2] };
	$options->{custom_inline_get}    = sub { my ($self, $inst) = @_; qq(\$MooseX::FunkyAttributes::Role::Attribute::InsideOut::_HASHES[$hashcount]{$inst}) };
	$options->{custom_inline_set}    = sub { my ($self, $inst, $val) = @_; qq(\$MooseX::FunkyAttributes::Role::Attribute::InsideOut::_HASHES[$hashcount]{$inst} = $val) };
	$options->{custom_inline_weaken} = sub { my ($self, $inst) = @_; qq(Scalar::Util::weaken \$MooseX::FunkyAttributes::Role::Attribute::InsideOut::_HASHES[$hashcount]{$inst}) };
	$options->{custom_inline_has}    = sub { my ($self, $inst) = @_; qq(exists \$MooseX::FunkyAttributes::Role::Attribute::InsideOut::_HASHES[$hashcount]{$inst}) };
	$options->{custom_inline_clear}  = sub { my ($self, $inst) = @_; qq(delete \$MooseX::FunkyAttributes::Role::Attribute::InsideOut::_HASHES[$hashcount]{$inst}) };
};

1;

__END__

=head1 NAME

MooseX::FunkyAttributes::Role::Attribute::InsideOut - an inside-out attribute

=head1 SYNOPSIS

   package Person;
   
   use Moose;
   use MooseX::FunkyAttributes;
   
   has name => (
      traits => [ InsideOutAttribute ],
      is     => 'ro',
      isa    => 'Str',
   );
   
   has age => (
      is     => 'ro',
      isa    => 'Num',
   );
   
   package main;
   
   use feature 'say';
   
   my $bob = Person->new(name => 'Bob', age => 32);
   say $bob->name;   # Bob
   say $bob->dump;   # $VAR1 = bless({ age => 32 }, 'Person');

=head1 DESCRIPTION

This trait implements the "inside-out" technique for Moose attributes. Unlike
L<MooseX::InsideOut> it doesn't make all attributes in the class inside-out;
just the attribute(s) it is applied to.

One situation where you might want to do this is to hide certain attributes
from dumps. For example, a "password" attribute that you don't want to appear
in log files, or an attribute which contains a large chunk of textual data or
a deeply nested data structure which makes the logs less readable.

This trait inherits from L<MooseX::FunkyAttributes::Role::Attribute>, but
forget about most of what you read in the documentaton for that trait. Those
C<custom_set>, C<custom_get>, C<custom_inline_set>, etc options are all
automatically generated for you.

=head1 BUGS

Please report any bugs to
L<http://rt.cpan.org/Dist/Display.html?Queue=MooseX-FunkyAttributes>.

=head1 SEE ALSO

L<MooseX::FunkyAttributes>, L<MooseX::InsideOut>, L<Hash::FieldHash>.

=head1 AUTHOR

Toby Inkster E<lt>tobyink@cpan.orgE<gt>.

=head1 COPYRIGHT AND LICENCE

This software is copyright (c) 2012-2013 by Toby Inkster.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=head1 DISCLAIMER OF WARRANTIES

THIS PACKAGE IS PROVIDED "AS IS" AND WITHOUT ANY EXPRESS OR IMPLIED
WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED WARRANTIES OF
MERCHANTIBILITY AND FITNESS FOR A PARTICULAR PURPOSE.



lib/MooseX/FunkyAttributes.pm
package MooseX::FunkyAttributes;

use 5.008;
use strict;
use warnings;

BEGIN {
	$MooseX::FunkyAttributes::AUTHORITY = 'cpan:TOBYINK';
	$MooseX::FunkyAttributes::VERSION   = '0.002';
}

use aliased qw(
	MooseX::FunkyAttributes::Role::Attribute
	FunkyAttribute
);

use aliased qw(
	MooseX::FunkyAttributes::Role::Attribute::InsideOut
	InsideOutAttribute
);

use aliased qw(
	MooseX::FunkyAttributes::Role::Attribute::Delegated
	DelegatedAttribute
);

use Exporter::Everything;

1;

__END__

=head1 NAME

MooseX::FunkyAttributes - add code smell to your Moose attributes

=head1 SYNOPSIS

   package Circle;
   
   use Moose;
   use MooseX::FunkyAttributes;
   
   has radius => (
      is          => 'rw',
      isa         => 'Num',
      predicate   => 'has_radius',
   );
   
   has diameter => (
      traits      => [ FunkyAttribute ],
      is          => 'rw',
      isa         => 'Num',
      custom_get  => sub { 2 * $_->radius },
      custom_set  => sub { $_->radius( $_[-1] / 2 ) },
      custom_has  => sub { $_->has_radius },
   );

=head1 DESCRIPTION

The MooseX::FunkyAttributes module itself just provides some convenience
functions for the attribute traits that are distributed with it. 

The grand uniting idea behind the traits is that although Moose generally
uses blessed hashrefs for object internals, storing each attribute as an
entry in the hash, sometimes you may want to change that behaviour for
individual attributes. For example, you could use the inside-out technique
for one attribute to suppress it from appearing in L<Data::Dumper> dumps.

The traits bundled with MooseX::FunkyAttributes are as follows. Please
see the documentation for each individual trait for further details.

=over

=item * L<MooseX::FunkyAttributes::Role::Attribute>

Allows you to override the get, set, clear and predicate operations for an
attribute. These are just the raw operations you're overriding; you do not
need to implement type constraint checking, coercians, checking required
attributes, defaults, builders, etc.

With this trait you can create attributes which are calculated on the fly
(as in the SYNOPSIS) or stored somewhere outside the object's blessed
hashref.

=item * L<MooseX::FunkyAttributes::Role::Attribute::InsideOut>

This trait stores the attribute using the inside-out technique. If you want
your whole object to be inside-out, then use L<MooseX::InsideOut>.

=item * L<MooseX::FunkyAttributes::Role::Attribute::Delegated>

This trait delegates the storage of one attribute to another attribute.
For example:

   package Head;
   use Moose;
   has mouth => (
      is           => 'ro',
      isa          => 'Mouth',
   );
   
   package Person;
   use Moose;
   use MooseX::FunkyAttributes;
   has head => (
      is           => 'ro',
      isa          => 'Head',
   );
   has mouth => (
      is           => 'ro',
      isa          => 'Mouth::Human',
      traits       => [ DelegatedAttribute ],
      delegated_to => 'head',
   );

It is not dissimilar to the Moose's idea of "handles" (in the example above,
we could have not defined a C<mouth> attribute as part of the C<Person> class,
and just specified that C<head> handles C<mouth>) however because C<mouth>
is now a proper attribute in the Person class (and will show up via Moose's
introspection API), it can have its own type constraints, coercion, trigger,
predicate, etc.

=back

The MooseX::FunkyAttributes module itself exports some constants (e.g.
C<FunkyAttribute> in the SYNOPSIS) which can be used as abbreviations for
the full trait package names. These constants are:

=over

=item C<FunkyAttribute>

=item C<InsideOutAttribute>

=item C<DelegatedAttribute>

=back

=head1 BUGS

Please report any bugs to
L<http://rt.cpan.org/Dist/Display.html?Queue=MooseX-FunkyAttributes>.

=head1 SEE ALSO

L<MooseX::FunkyAttributes::Role::Attribute>,
L<MooseX::FunkyAttributes::Role::Attribute::InsideOut>,
L<MooseX::FunkyAttributes::Role::Attribute::Delegated>.

These effect storage for whole object instances; not just one attribute:
L<MooseX::GlobRef>,
L<MooseX::InsideOut>,
L<MooseX::ArrayRef>.

L<MooseX::CustomInitArgs> - if you have (as in the SYNOPSIS) one attribute
which is calculated from another (diameter from radius), MooseX::CustomInitArgs
will allow you to accept both attributes in the constructor (i.e. accept a
diameter in the constructor, halve it, and set the radius attribute).

=head1 AUTHOR

Toby Inkster E<lt>tobyink@cpan.orgE<gt>.

=head1 COPYRIGHT AND LICENCE

This software is copyright (c) 2012-2013 by Toby Inkster.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=head1 DISCLAIMER OF WARRANTIES

THIS PACKAGE IS PROVIDED "AS IS" AND WITHOUT ANY EXPRESS OR IMPLIED
WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED WARRANTIES OF
MERCHANTIBILITY AND FITNESS FOR A PARTICULAR PURPOSE.

